Frameworks/UIKit/UIViewController.mm

```diff
@@ -1211,57 +1211,71 @@ - (void)_dismissTransitionStopped:(id)anim finished:(BOOL)finished {
     [[UIApplication sharedApplication] endIgnoringInteractionEvents];
 }
 
+- (void)_childDismissCleanup {
+    priv->_presentedViewController->priv->_parentViewController = nil;
+    priv->_presentedViewController->priv->_presentingViewController = nil;
+    priv->_presentedViewController->priv->_popoverPresentationController = nil;
+    priv->_presentedViewController = nil;
+}
+
 /**
  @Status Interoperable
 */
 - (void)dismissViewControllerAnimated:(BOOL)animated completion:(void (^)(void))completion {
-    if (priv->_modalViewController == nil) {
-        if ([self parentViewController] != nil) {
+    if (![self presentedViewController]) {
+        // Calling dismiss on a view controller that hasn't presented additional view controllers itself should result in the parent handling the dismissal.
+
+        if ([self parentViewController]) {
             [[self parentViewController] dismissViewControllerAnimated:animated completion:completion];
-            return;
+        } else {
+            TraceWarning(TAG, L"dismissViewController invalid - nothing to dismiss");
         }
 
-        TraceWarning(TAG, L"dismissModalViewController invalid!");
         return;
     }
 
-    UIViewController* curController = priv->_modalViewController;
+    UIViewController* presented = [self presentedViewController];
+
+    if ([presented presentedViewController]) {
+        // Recursively handle multiple child dismissal.
+        // TODO: This is unsafe because we end up dispatching the dimiss animation of a
+        // parent before the child dispatch animation completes.
```

Comment:
I've got some changes for handling the dismissal in the required order that I can handle in a third commit to this pull.